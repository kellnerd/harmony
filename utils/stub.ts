import { urlToFilePath } from './file_path.ts';
import { dirname } from '@std/path/dirname';
import { stub } from '@std/testing/mock';

/** CLI flag (`--download`) which allows {@linkcode stubFetchWithCache} to make network requests. */
export const downloadMode = Deno.args.includes('--download');

/**
 * Stubs {@linkcode fetch} to load the response from a cache instead of making a network request.
 *
 * Only in {@linkcode downloadMode} the resource will be fetched and written to the cache (as a file).
 */
export function stubFetchWithCache(cacheDir = 'testdata') {
	const originalFetch = globalThis.fetch;

	return stub(globalThis, 'fetch', async function (input, init) {
		const url = new URL(input instanceof Request ? input.url : input);
		const path = await urlToFilePath(url, { baseDir: cacheDir });

		if (downloadMode) {
			const response = await originalFetch(input, init);
			const { body } = response.clone();
			if (response.ok && body) {
				await Deno.mkdir(dirname(path), { recursive: true });
				await Deno.writeFile(path, body);
			}
			return response;
		} else {
			try {
				const body = await Deno.readFile(path);
				return new Response(body);
			} catch (error) {
				if (error instanceof Deno.errors.NotFound) {
					throw new Error(`Response for ${url} has not been cached`);
				} else {
					throw error;
				}
			}
		}
	});
}
