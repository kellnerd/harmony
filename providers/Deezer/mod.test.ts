import type { ReleaseOptions } from '@/harmonizer/types.ts';
import { describeProvider, makeProviderOptions } from '@/providers/test_spec.ts';
import { stubFetchWithCache } from '@/utils/stub.ts';
import { assert } from 'std/assert/assert.ts';
import { describe } from '@std/testing/bdd';

import DeezerProvider from './mod.ts';

describe('Deezer provider', () => {
	using _fetchStub = stubFetchWithCache();
	const deezer = new DeezerProvider(makeProviderOptions());

	// Standard options which have an effect for Deezer.
	const releaseOptions: ReleaseOptions = {
		withSeparateMedia: true,
		withAllTrackArtists: true,
		withISRC: true,
	};

	describeProvider(deezer, {
		urls: [{
			description: 'release page',
			url: new URL('https://www.deezer.com/album/11591214'),
			id: { type: 'album', id: '11591214' },
			isCanonical: true,
		}, {
			description: 'localized release page',
			url: new URL('https://www.deezer.com/fr/album/629506181'),
			id: { type: 'album', id: '629506181' },
		}, {
			description: 'localized release page without www subdomain',
			url: new URL('https://deezer.com/en/album/521038732'),
			id: { type: 'album', id: '521038732' },
		}, {
			description: 'artist page',
			url: new URL('https://www.deezer.com/artist/8686'),
			id: { type: 'artist', id: '8686' },
			isCanonical: true,
		}, {
			description: 'playlist page',
			url: new URL('https://www.deezer.com/en/playlist/1976454162'),
		}],
		releaseLookup: [{
			description: 'single by two artists',
			release: new URL('https://www.deezer.com/en/album/629506181'),
			options: releaseOptions,
			assert: (release) => {
				const allTracks = release.media.flatMap((medium) => medium.tracklist);
				assert(allTracks[0].artists?.length === 2, 'Main track should have two artists');
				assert(allTracks.every((track) => track.isrc), 'All tracks should have an ISRC');
			},
		}, {
			description: 'single by two artists (without additional lookup options)',
			release: '629506181', // same single as the previous one
			skipSnapshot: true, // just a subset of the previous one
			assert: (release) => {
				const allTracks = release.media.flatMap((medium) => medium.tracklist);
				assert(allTracks.every((track) => track.artists?.length === 1), 'Tracks should not have multiple artists');
				assert(allTracks.every((track) => !track.isrc), 'Tracks should not have an ISRC');
			},
		}],
	});
});
