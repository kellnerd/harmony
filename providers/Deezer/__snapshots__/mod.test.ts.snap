export const snapshot = {};

snapshot[`Deezer provider > release lookup > single by two artists 1`] = `
{
  artists: [
    {
      creditedName: "Lady Gaga",
      externalIds: [
        {
          id: "75491",
          provider: "deezer",
          type: "artist",
        },
      ],
      name: "Lady Gaga",
    },
    {
      creditedName: "Bruno Mars",
      externalIds: [
        {
          id: "429675",
          provider: "deezer",
          type: "artist",
        },
      ],
      name: "Bruno Mars",
    },
  ],
  availableIn: [
    "AE",
    "AF",
    "AG",
    "AI",
    "AL",
    "AM",
    "AO",
    "AR",
    "AT",
    "AU",
    "AZ",
    "BA",
    "BB",
    "BD",
    "BE",
    "BF",
    "BG",
    "BH",
    "BI",
    "BJ",
    "BN",
    "BO",
    "BR",
    "BT",
    "BW",
    "CA",
    "CC",
    "CD",
    "CF",
    "CG",
    "CH",
    "CI",
    "CK",
    "CL",
    "CM",
    "CO",
    "CR",
    "CV",
    "CX",
    "CY",
    "CZ",
    "DE",
    "DJ",
    "DK",
    "DM",
    "DO",
    "DZ",
    "EC",
    "EE",
    "EG",
    "ER",
    "ES",
    "ET",
    "FI",
    "FJ",
    "FK",
    "FM",
    "FR",
    "GA",
    "GB",
    "GD",
    "GE",
    "GH",
    "GM",
    "GN",
    "GQ",
    "GR",
    "GT",
    "GW",
    "HN",
    "HR",
    "HU",
    "ID",
    "IE",
    "IL",
    "IQ",
    "IS",
    "IT",
    "JM",
    "JO",
    "JP",
    "KE",
    "KG",
    "KH",
    "KI",
    "KM",
    "KN",
    "KW",
    "KY",
    "KZ",
    "LA",
    "LB",
    "LC",
    "LK",
    "LR",
    "LS",
    "LT",
    "LU",
    "LV",
    "LY",
    "MA",
    "MD",
    "ME",
    "MG",
    "MH",
    "MK",
    "ML",
    "MN",
    "MR",
    "MS",
    "MT",
    "MU",
    "MV",
    "MW",
    "MX",
    "MY",
    "MZ",
    "NA",
    "NE",
    "NF",
    "NG",
    "NI",
    "NL",
    "NO",
    "NP",
    "NR",
    "NU",
    "NZ",
    "OM",
    "PA",
    "PE",
    "PG",
    "PH",
    "PK",
    "PL",
    "PN",
    "PT",
    "PW",
    "PY",
    "QA",
    "RO",
    "RS",
    "RW",
    "SA",
    "SB",
    "SC",
    "SE",
    "SG",
    "SI",
    "SJ",
    "SK",
    "SL",
    "SN",
    "SO",
    "ST",
    "SV",
    "SZ",
    "TC",
    "TD",
    "TG",
    "TH",
    "TJ",
    "TK",
    "TL",
    "TM",
    "TN",
    "TO",
    "TR",
    "TV",
    "TZ",
    "UA",
    "UG",
    "US",
    "UY",
    "UZ",
    "VC",
    "VE",
    "VG",
    "VN",
    "VU",
    "WS",
    "YE",
    "ZA",
    "ZM",
    "ZW",
  ],
  excludedFrom: [
    "IO",
  ],
  externalLinks: [
    {
      types: [
        "free streaming",
      ],
      url: URL {
        hash: "",
        host: "www.deezer.com",
        hostname: "www.deezer.com",
        href: "https://www.deezer.com/album/629506181",
        origin: "https://www.deezer.com",
        password: "",
        pathname: "/album/629506181",
        port: "",
        protocol: "https:",
        search: "",
        username: "",
      },
    },
  ],
  gtin: "602475093060",
  images: [
    {
      thumbUrl: URL {
        hash: "",
        host: "cdn-images.dzcdn.net",
        hostname: "cdn-images.dzcdn.net",
        href: "https://cdn-images.dzcdn.net/images/cover/4bd5903f4ce8f2601916bfadb44efe8a/250x250-000000-80-0-0.jpg",
        origin: "https://cdn-images.dzcdn.net",
        password: "",
        pathname: "/images/cover/4bd5903f4ce8f2601916bfadb44efe8a/250x250-000000-80-0-0.jpg",
        port: "",
        protocol: "https:",
        search: "",
        username: "",
      },
      types: [
        "front",
      ],
      url: URL {
        hash: "",
        host: "cdn-images.dzcdn.net",
        hostname: "cdn-images.dzcdn.net",
        href: "https://cdn-images.dzcdn.net/images/cover/4bd5903f4ce8f2601916bfadb44efe8a/1000x1000-000000-80-0-0.jpg",
        origin: "https://cdn-images.dzcdn.net",
        password: "",
        pathname: "/images/cover/4bd5903f4ce8f2601916bfadb44efe8a/1000x1000-000000-80-0-0.jpg",
        port: "",
        protocol: "https:",
        search: "",
        username: "",
      },
    },
  ],
  info: {
    messages: [],
    providers: [
      {
        apiUrl: URL {
          hash: "",
          host: "api.deezer.com",
          hostname: "api.deezer.com",
          href: "https://api.deezer.com/album/629506181",
          origin: "https://api.deezer.com",
          password: "",
          pathname: "/album/629506181",
          port: "",
          protocol: "https:",
          search: "",
          username: "",
        },
        id: "629506181",
        internalName: "deezer",
        lookup: {
          method: "id",
          value: "629506181",
        },
        name: "Deezer",
        url: URL {
          hash: "",
          host: "www.deezer.com",
          hostname: "www.deezer.com",
          href: "https://www.deezer.com/album/629506181",
          origin: "https://www.deezer.com",
          password: "",
          pathname: "/album/629506181",
          port: "",
          protocol: "https:",
          search: "",
          username: "",
        },
      },
    ],
  },
  labels: [
    {
      name: "Interscope",
    },
  ],
  media: [
    {
      format: "Digital Media",
      number: 1,
      tracklist: [
        {
          artists: [
            {
              creditedName: "Lady Gaga",
              externalIds: [
                {
                  id: "75491",
                  provider: "deezer",
                  type: "artist",
                },
              ],
              name: "Lady Gaga",
            },
            {
              creditedName: "Bruno Mars",
              externalIds: [
                {
                  id: "429675",
                  provider: "deezer",
                  type: "artist",
                },
              ],
              name: "Bruno Mars",
            },
          ],
          availableIn: [
            "AE",
            "AF",
            "AG",
            "AI",
            "AL",
            "AM",
            "AO",
            "AQ",
            "AR",
            "AS",
            "AT",
            "AU",
            "AZ",
            "BA",
            "BB",
            "BD",
            "BE",
            "BF",
            "BG",
            "BH",
            "BI",
            "BJ",
            "BN",
            "BO",
            "BQ",
            "BR",
            "BT",
            "BV",
            "BW",
            "CA",
            "CC",
            "CD",
            "CF",
            "CG",
            "CH",
            "CI",
            "CK",
            "CL",
            "CM",
            "CO",
            "CR",
            "CU",
            "CV",
            "CW",
            "CX",
            "CY",
            "CZ",
            "DE",
            "DJ",
            "DK",
            "DM",
            "DO",
            "DZ",
            "EC",
            "EE",
            "EG",
            "EH",
            "ER",
            "ES",
            "ET",
            "FI",
            "FJ",
            "FK",
            "FM",
            "FR",
            "GA",
            "GB",
            "GD",
            "GE",
            "GH",
            "GM",
            "GN",
            "GQ",
            "GR",
            "GS",
            "GT",
            "GU",
            "GW",
            "HK",
            "HM",
            "HN",
            "HR",
            "HU",
            "ID",
            "IE",
            "IL",
            "IN",
            "IQ",
            "IR",
            "IS",
            "IT",
            "JM",
            "JO",
            "JP",
            "KE",
            "KG",
            "KH",
            "KI",
            "KM",
            "KN",
            "KP",
            "KR",
            "KW",
            "KY",
            "KZ",
            "LA",
            "LB",
            "LC",
            "LK",
            "LR",
            "LS",
            "LT",
            "LU",
            "LV",
            "LY",
            "MA",
            "MD",
            "ME",
            "MG",
            "MH",
            "MK",
            "ML",
            "MM",
            "MN",
            "MP",
            "MR",
            "MS",
            "MT",
            "MU",
            "MV",
            "MW",
            "MX",
            "MY",
            "MZ",
            "NA",
            "NE",
            "NF",
            "NG",
            "NI",
            "NL",
            "NO",
            "NP",
            "NR",
            "NU",
            "NZ",
            "OM",
            "PA",
            "PE",
            "PG",
            "PH",
            "PK",
            "PL",
            "PN",
            "PS",
            "PT",
            "PW",
            "PY",
            "QA",
            "RO",
            "RS",
            "RW",
            "SA",
            "SB",
            "SC",
            "SD",
            "SE",
            "SG",
            "SI",
            "SJ",
            "SK",
            "SL",
            "SN",
            "SO",
            "ST",
            "SV",
            "SX",
            "SY",
            "SZ",
            "TC",
            "TD",
            "TG",
            "TH",
            "TJ",
            "TK",
            "TL",
            "TM",
            "TN",
            "TO",
            "TR",
            "TV",
            "TZ",
            "UA",
            "UG",
            "US",
            "UY",
            "UZ",
            "VC",
            "VE",
            "VG",
            "VI",
            "VN",
            "VU",
            "WS",
            "YE",
            "ZA",
            "ZM",
            "ZW",
          ],
          isrc: "USUM72409273",
          length: 250000,
          number: 1,
          title: "Die With A Smile",
        },
      ],
    },
  ],
  packaging: "None",
  releaseDate: {
    day: 16,
    month: 8,
    year: 2024,
  },
  status: "Official",
  title: "Die With A Smile",
  types: [
    "Single",
  ],
}
`;
